crates/qpy/src/circuits.rs

```diff
@@ -0,0 +1,2174 @@
+// 3. "Write": To write to a file obj the serialization of the original data
+// Ideally, serialization is done by packing in a binrw-enhanced struct and using the
+// `write` method into a `Cursor` buffer, but there might be exceptions.
+
+use hashbrown::HashMap;
+use numpy::IntoPyArray;
+use numpy::ToPyArray;
+use pyo3::exceptions::{PyIOError, PyValueError};
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::IntoPyDict;
+use pyo3::types::{PyAny, PyBytes, PyDict, PyIterator, PyList, PyString, PyTuple, PyType};
+use pyo3::IntoPyObjectExt;
+
+use qiskit_circuit::bit::{
+    ClassicalRegister, PyClassicalRegister, PyClbit, QuantumRegister, Register, ShareableClbit,
+    ShareableQubit,
+};
+use qiskit_circuit::circuit_data::{CircuitData, CircuitStretchType, CircuitVarType};
+use qiskit_circuit::circuit_instruction::{CircuitInstruction, OperationFromPython};
+use qiskit_circuit::classical;
+use qiskit_circuit::imports;
+use qiskit_circuit::operations::{ArrayType, Operation, OperationRef, Param, StandardInstruction};
+use qiskit_circuit::packed_instruction::{PackedInstruction, PackedOperation};
+use qiskit_circuit::{Clbit, Qubit};
+
+use smallvec::SmallVec;
+use uuid::Uuid;
+
+use crate::annotations::AnnotationHandler;
+use crate::bytes::Bytes;
+use crate::consts::standard_gate_from_gate_class_name;
+use crate::formats;
+use crate::params::{pack_param, unpack_param};
+use crate::value::{
+    circuit_instruction_types, deserialize, deserialize_with_args, dumps_value,
+    expression_var_declaration, get_circuit_type_key, pack_generic_data, pack_standalone_var,
+    serialize, tags, unpack_generic_data, DumpedValue, ExpressionType, QPYReadData, QPYWriteData,
+};
+
+use crate::UnsupportedFeatureForVersion;
+
+const UNITARY_GATE_CLASS_NAME: &str = "UnitaryGate";
+// used when writing
+type CustomOperationsMap = HashMap<String, PackedOperation>;
+type CustomOperationsList = Vec<String>;
+//used when reading
+#[derive(Debug)]
+struct CustomCircuitInstructionData {
+    gate_type: u8,
+    num_qubits: u32,
+    num_clbits: u32,
+    definition_circuit: Option<Py<PyAny>>,
+    // num_ctrl_qubits: u32,
+    // ctrl_state: u32,
+    base_gate_raw: Bytes,
+}
+
+type CustomInstructionsMap = HashMap<String, CustomCircuitInstructionData>;
+
+pub mod register_types {
+    pub const QREG: u8 = b'q';
+    pub const CREG: u8 = b'c';
+}
+
+fn getattr_or_none<'py>(
+    py_object: &'py Bound<'py, PyAny>,
+    name: &str,
+) -> PyResult<Option<Bound<'py, PyAny>>> {
+    if py_object.hasattr(name)? {
+        let attr = py_object.getattr(name)?;
+        if attr.is_none() {
+            Ok(None)
+        } else {
+            Ok(Some(attr))
+        }
+    } else {
+        Ok(None)
+    }
+}
+fn is_python_gate(py: Python, op: &PackedOperation, python_gate: &Bound<PyAny>) -> PyResult<bool> {
+    match op.view() {
+        OperationRef::Gate(pygate) => {
+            if pygate.gate.bind(py).is_instance(python_gate)? {
```

Comment:
I think you might be able to check if the instance is the same even without binding.